src/com/cdd/bao/validator/RemappingChecker.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * BioAssay Ontology Annotator Tools
+ *
+ * (c) 2014-2018 Collaborative Drug Discovery Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License 2.0
+ * as published by the Free Software Foundation:
+ *
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package com.cdd.bao.validator;
+
+import java.io.*;
+import java.util.*;
+
+/*
+	Use logic in this class to detect cycles in remapping directives. 
+*/
+
+public class RemappingChecker
+{
+	public static void validateRemappings(Map<String, String> remappings) throws IOException
+	{
+		List<String> dependencies = new ArrayList<>();
```

Comment:
In case, the original comment gets lost due to the update.

This needs to be a LinkedHashSet. The add method of a Set returns true if the element was added and false if not (because it was already an element of the set). List.add probably always allows adding an element. This method needs to have a test.

Reply:
I think I'm just going to revert to the original logic.  Linked lists generally have inferior performance to lists.  But that aside, I think we are buying concision at the expense of simplicity.... so unless there's an objection, I'm fully reverting to the original logic.

Reply:
I suggested the LinkedHashSet as you were using the dependencies in the Exception. If you don't care about the order, a HashSet would be sufficient.

Reply:
BTW, [here](https://www.youtube.com/watch?v=YQs6IC-vgmo) is Stroustrup's talk on linked lists and why it's advisable to avoid them.  In our case, clearly there's little performance bonus for either lists or linked lists since our remapping chains in all likelihood will be tiny...  Still, I'm opting for lists.

Reply:
Regarding order:  my original thinking is that if we ever did see an exception, it might be nice to see the chain of remappings in sequential order.... hence, the `List`.

Reply:
I think the List implementation looks good. The higher order datastructures usually have enough overhead to cancel out the benefits when the size is capped at single figures... and keeping the original order is almost always a good thing when user feedback is happening. (Though do keep in mind that Jena's triple reader already randomises everything, so in some cases it's already lost...)

Reply:
Also, TreeSet/LinkedHashSet probably should've been named by function rather than implementation, e.g. SortedSet/OriginalOrderSet, but I've sorta gotten used to them.

Reply:
Agree on that one ... tend to forget.